framework/doc/content/finite_volumes/linear_fv_design.md

```diff
@@ -0,0 +1,167 @@
+
+### Elemental Kernels
+
+For volumetric effects, an element-based evaluation is necessary. Good examples are
+external source terms or reaction terms in the partial differential equations.
+To manage the contributions of these terms to the linear system, a
+`LinearFVElementalKernel` base class has been added which can be used for inheritance.
+The derived kernels need to override the following functions:
+
+- `computeMatrixContribution` which computes the matrix contributions to the degree of
+  freedom corresponding to the element.
+- `computeRightHandSideContribution` which computes the right hand side contributions
+  to the degree of freedom corresponding to the element.
+
+
+The population of the system matrix and right hand side with face terms
+(from flux kernels) is done in a face loop: `ComputeLinearFVFaceThread`.
+
+On the other hand, the population of the system matrix and right hand side
+with element-based volumetric terms is done in an element
+loop: `ComputeLinearFVElementalThread`.
+
+## Gradient Computation
+
+Considering that the linear finite volume system is used in a Picard-style iteration,
+terms that need cell gradients are lagged. Treating gradient-based terms explicitly results
+in faster linear solves due to the matrix containing fewer entries. Such terms include:
+
+- The linear terms in the extrapolated boundary conditions.
+- The nonorthogonal correctors in the surface normal gradient computation.
+
+The explicit treatment of these terms allows us to keep only one ghosted layer
+(of elements shared between processes) in parallel simulations, minimizing the
+linear solve time and parallel communication
+time over an iteration. The setbacks of this design choice are:
+
+- We need corrector iterations to resolve extrapolated boundary conditions
+  and nonorthogonal correctors. This means that even linear systems may need to be solved
+  multiple times to get a converged solution.
+- It may increase the number of iterations needed to resolve nonlinearities.
+
+## Boundary Conditions
+
+A significant difference compared to other systems in MOOSE is the way
+boundary conditions are enforced. Even though the user is still required to define the
+boundary conditions in the `[LinearFVBCs]` block, these boundary conditions
+are not executed on their own; instead they are used within `LinearFVFluxKernel`-s,
+through the following two functions:
+
+- `computeBoundaryMatrixContribution` which computes the matrix contributions of
+  a boundary face to the degree of freedom corresponding to the adjacent element.
+- `computeBoundaryRHSContribution` which computes the right hand side contributions of
+  a boundary face to the degree of freedom corresponding to the adjacent element.
+
+To create a new boundary condition the following functions need to be overridden:
+
+- `computeBoundaryValue` computes the boundary value of the field.
+- `computBoundaryNormalGradient` computes the normal gradient of the variable on this boundary.
```

Comment:
I don't know if I agree. `LinearFVFunctorDirichletBC` seems pretty generic and independent of whatever physics/kernels are present

Reply:
The matrix and the RHS terms are not going to be the same regardless of all kernels. This is undeniaby true.
The advection and diffusion kernel dont even use the same APIs from the BCs, and an interface reaction term would use another API as well. 

If we had a curl term which got integrated by parts instead of a diffusion or advection term, you'd need to add the computeBoundaryCurlRHS/Matrix APIs. For any arbitrary form kernel, you ll have to either had a new API or correct the output of the existing ones. This is not the kind of generality we have with a DirichletBC in NLFV or CGFE.

Reply:
I'll rework this in the next commit.